enterprise/server/workflow/service/service.go

```diff
@@ -610,25 +561,104 @@ func (ws *workflowService) getActions(ctx context.Context, wf *tables.Workflow,
+			`SELECT * FROM Workflows WHERE workflow_id = ?`,
+			workflowID,
+		).Take(wf).Error
+		if err != nil {
+			if db.IsRecordNotFound(err) {
+				return nil, status.NotFoundError("Workflow not found")
+			}
+			return nil, status.InternalError(err.Error())
+		}
+		return wf, nil
 	}
-	groupID, repoURL, err := parseRepositoryWorkflowID(id)
+
+	// If the workflow ID identifies a GitRepository, look up the GitRepository and construct a synthetic Workflow
+	// from it.
+	groupID, repoURL, err := parseRepositoryWorkflowID(workflowID)
+	rwf, err := ws.getRepositoryWorkflow(ctx, groupID, repoURL)
 	if err != nil {
 		return nil, err
 	}
+
+	return rwf.Workflow, nil
+}
+
+func (ws *workflowService) GetWorkflowByGroupAndRepo(ctx context.Context, groupID string, repoURL string) (*tables.Workflow, error) {
+	parsedRepoURL, err := gitutil.ParseGitHubRepoURL(repoURL)
+	if err != nil {
+		return nil, err
+	}
+
+	rwf, err := ws.getRepositoryWorkflow(ctx, groupID, parsedRepoURL)
+	if err == nil {
+		return rwf.Workflow, nil
+	}
+
+	// If the workflow isn't found in the GitRepositories table, look in the legacy Workflows table
+	wf := &tables.Workflow{}
+	err = ws.env.GetDBHandle().DB(ctx).Raw(`
+		SELECT *
+		FROM Workflows
+		WHERE group_id = ?
+		AND repo_url = ?
+	`, groupID, repoURL).Take(wf).Error
+	if err != nil {
+		if db.IsRecordNotFound(err) {
+			return nil, status.NotFoundErrorf("repo %q not found", repoURL)
+		}
+		return nil, status.InternalErrorf("failed to look up repo %q: %s", repoURL, err)
+	}
+
+	return wf, nil
+}
+
+// To run workflow in a clean container, update the instance name suffix
+func (ws *workflowService) useCleanWorkflow(ctx context.Context, wf *tables.Workflow) error {
+	suffix, err := random.RandomString(10)
+	if err != nil {
+		return err
+	}
+	wf.InstanceNameSuffix = suffix
+
+	if isRepositoryWorkflowID(wf.WorkflowID) {
+		err = ws.env.GetDBHandle().DB(ctx).Exec(`
+				UPDATE GitRepositories
+				SET instance_name_suffix = ?
+				WHERE group_id = ? AND repo_url = ?`,
+			suffix, wf.GroupID, wf.RepoURL,
+		).Error
+	} else {
+		err = ws.env.GetDBHandle().DB(ctx).Exec(`
+				UPDATE Workflows
+				SET instance_name_suffix = ?
+				WHERE workflow_id = ?`,
+			wf.InstanceNameSuffix, wf.WorkflowID,
+		).Error
+	}
+	return err
+}
+
+func (ws *workflowService) getRepositoryWorkflow(ctx context.Context, groupID string, repoURL *gitutil.RepoURL) (*repositoryWorkflow, error) {
+	app := ws.env.GetGitHubApp()
+	if app == nil {
+		return nil, status.UnimplementedError("GitHub App is not configured")
+	}
 	if err := perms.AuthorizeGroupAccess(ctx, ws.env, groupID); err != nil {
 		return nil, err
 	}
 	gitRepository := &tables.GitRepository{}
-	err = ws.env.GetDBHandle().DB(ctx).Raw(`
+	err := ws.env.GetDBHandle().DB(ctx).Raw(`
 		SELECT *
 		FROM GitRepositories
 		WHERE group_id = ?
 		AND repo_url = ?
-	`, groupID, repoURL.String()).Take(gitRepository).Error
+	`, groupID, repoURL).Take(gitRepository).Error
```

Comment:
not sure if GORM knows to automatically call `.String()` here, but it feels a bit less "magical" to be explicit about it either way:

```suggestion
	`, groupID, repoURL.String()).Take(gitRepository).Error
```